File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/digitaltwin/DigitalTwinSpace.java
package org.opcfoundation.webserver.digitaltwin;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.opcfoundation.webserver.addressspace.nodes.UaEnumDataType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerReactiveObject;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaObjects;
import org.opcfoundation.webserver.digitaltwin.digitaltwin.DigitalTwinRepositoryType;
import org.opcfoundation.webserver.digitaltwin.digitaltwin.DigitalTwinType;
import org.opcfoundation.webserver.digitaltwin.element.ElementType;
import org.opcfoundation.webserver.digitaltwin.submodel.SubmodelTypeBase;
import org.opcfoundation.webserver.types.common.UaInstanceIdentifier;
import org.opcfoundation.webserver.types.common.UaObjectIdentifier;

public class DigitalTwinSpace extends NodeManagerReactiveObject {
    public DigitalTwinSpace(String namespaceUri)
    {
        super(namespaceUri);
    }

    public final void addDefinition(DigitalTwinType type)
    {
        addObjectType(type);
    }

    public final void addDefinition(DigitalTwinRepositoryType type)
    {
        addObjectType(type);
    }

    public final void addDefinition(SubmodelTypeBase type)
    {
        addObjectType(type);
    }

    public final void addDefinition(ElementType type)
    {
        addObjectType(type);
    }

    public final void addDefinition(UaEnumDataType type)
    {
        addDataType(type);
    }

    public final void addRepository(
            DigitalTwinRepositoryType repositoryType,
            String id,
            LocalizedText displayName,
            LocalizedText description)
    {
        UaInstanceIdentifier objectIdentifier = new UaInstanceIdentifier(
                new UaObjectIdentifier(repositoryType.nodeId().toParseableString(), id, null),
                null);

        NodeId objectNodeId = new NodeId(nsIndex(), objectIdentifier.toByteString());

        UaObject newObject = new UaObject(
                objectNodeId,
                id,
                displayName,
                repositoryType);

        newObject.setDescription(description);

        this.addNode(newObject);

        UaObjects.ObjectsFolder.organizes(newObject);
    }
}

File: opcua-webservice-node/src/server/digitaltwin/DigitalTwinSpace.ts
import { UaLocalizedText, UaNodeId } from "opcua-webapi-ts";
import { NodeManagerReactiveObject } from "../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaObject } from "../addressspace/nodes/UaObject";
import { UaObjects } from "../addressspace/nodes/builtin/UaObjects";
import { UaInstanceIdentifier, UaObjectIdentifier } from "../types";
import type { DigitalTwinRepositoryType } from "./digitaltwin/DigitalTwinRepositoryType";

export class DigitalTwinSpace extends NodeManagerReactiveObject {
    constructor(namespaceUri: string) {
        super(namespaceUri);
    }

    addRepository(
        repositoryType: DigitalTwinRepositoryType,
        id: string,
        displayName: UaLocalizedText,
        description: UaLocalizedText,
    ): void {
        const objectIdentifier = new UaInstanceIdentifier(
            new UaObjectIdentifier(repositoryType.nodeId.toString(), id, null),
            null,
        );

        const objectNodeId = new UaNodeId(objectIdentifier.toByteString(), this.nsIndex());
        const newObject = new UaObject(objectNodeId, id, displayName, repositoryType);
        newObject.description = description;

        this.addNode(newObject);
        UaObjects.ObjectsFolder.organizes(newObject);
    }
}
